[posix] allow set POSIX TUN device at runtime

wtongze opened the pull request:
This commit introduces allows users to pass the POSIX TUN device path through a command line flag (`--tun-device`) when starting the daemon / cli if `OT_PLATFORM_NETIF` is enabled.
- If users doesn't not set this flag, the POSIX TUN device path will be default to `OPENTHREAD_POSIX_TUN_DEVICE`.

wtongze commented:
/gemini review

yangsong-cnyn commented:
@wtongze please use https://github.com/openthread/openthread/blob/4f38928041b2207b33ab2bb69046c4721907aae5/script/make-pretty to format the files.

wtongze commented on src/posix/platform/include/openthread/openthread-system.h:
```diff
@@ -83,6 +83,9 @@ typedef struct otPlatformCoprocessorUrls
  */
 typedef struct otPlatformConfig
 {
+#if OPENTHREAD_POSIX_TUN_DEVICE_SET_API_ENABLE
```
Thanks for pointing it out. I will remove the `#if` check here and add a comment.

bukepo commented on src/posix/main.c:
```diff
@@ -161,6 +167,9 @@ static void PrintUsage(const char *aProgramName, FILE *aStream, int aExitCode)
             "Options:\n"
+#if OPENTHREAD_POSIX_CONFIG_TUN_DEVICE_ENABLE
+            "        --tun-device              POSIX TUN Device.\n"
+#endif
```
```suggestion
            "        --tun-device              POSIX TUN Device.\n"
```

bukepo commented on src/posix/platform/netif.cpp:
```diff
@@ -2033,7 +2018,12 @@ static void platformConfigureTunDevice(otPlatformConfig *aPlatformConfig)
     sTunFd = open(OPENTHREAD_POSIX_TUN_DEVICE, O_RDWR | O_CLOEXEC | O_NONBLOCK);
+#else
+    sTunFd = open(aPlatformConfig->mTunDevice, O_RDWR | O_CLOEXEC | O_NONBLOCK);
+#endif
```
```suggestion
    sTunFd = open(aPlatformConfig->mTunDevice, O_RDWR | O_CLOEXEC | O_NONBLOCK);
```

bukepo commented on src/posix/platform/openthread-posix-config.h:
```diff
@@ -484,6 +484,40 @@
+#ifndef OPENTHREAD_POSIX_CONFIG_TUN_DEVICE_ENABLE
+#define OPENTHREAD_POSIX_CONFIG_TUN_DEVICE_ENABLE 0
+#endif
+
```
```suggestion
```

wtongze committed:
[posix] allow set POSIX TUN device at runtime
This commit allows users to pass the POSIX TUN device path
through a command line flag when starting the daemon / cli
when `OT_PLATFORM_NETIF` is enabled.

zhanglongxia reviewed (approved):
LGTM

bukepo commented on src/posix/platform/netif.cpp:
```diff
@@ -2033,7 +2033,8 @@ static void platformConfigureTunDevice(otPlatformConfig *aPlatformConfig)
     const char  *interfaceName;
 
-    sTunFd = open(OPENTHREAD_POSIX_TUN_DEVICE, O_RDWR | O_CLOEXEC | O_NONBLOCK);
+    sTunFd = open((aPlatformConfig->mTunDevice == NULL ? OPENTHREAD_POSIX_TUN_DEVICE : aPlatformConfig->mTunDevice),
```
Or initialize `aPlatformConfig->mTunDevice` in `platformNetifInit()`.

bukepo replied:
it would be simpler if initializing in `platformNetifInit()`. No extra function is needed.

wtongze replied:
> it would be simpler if initializing in `platformNetifInit()`. No extra function is needed.

Inside `platformNetifInit` it actually calls platform-specific `platformConfigureTunDevice` function to initialize `sTunFd`.

wtongze committed:
refactor: wrap to platformGetTunDevicePath function

yangsong-cnyn reviewed (approved):
LGTM, thanks!

wtongze committed:
refactor: remove  function

wtongze committed:
fix: add platform check

wtongze committed:
refactor: change to nullptr

zhanglongxia commented on src/posix/main.c:
```diff
@@ -189,6 +198,9 @@ static void ParseArg(int aArgCount, char *aArgVector[], PosixConfig *aConfig)
     aConfig->mLogLevel                            = OT_LOG_LEVEL_CRIT;
     aConfig->mPlatformConfig.mInterfaceName       = OPENTHREAD_POSIX_CONFIG_THREAD_NETIF_DEFAULT_NAME;
+#if OPENTHREAD_CONFIG_PLATFORM_NETIF_ENABLE
+    aConfig->mPlatformConfig.mTunDevice = NULL;
```
Should we set `aConfig->mPlatformConfig.mTunDevice = OPENTHREAD_POSIX_TUN_DEVICE ` here to avoid checking whether the `aConfig->mPlatformConfig.mTunDevice == nullptr` before using the `aConfig->mPlatformConfig.mTunDevice`?

zhanglongxia reviewed (approved):
LGTM

abtink reviewed (approved):
LGTM.
